### PR TITLE
Moving Averages

### PR DESCRIPTION
**Pull request by krishnadurai**
This PR implements MovingAverages of temperatures across a number of measurements determined by a constant 'averagingWindow'.

An average is determined inclusive of the current measurement and the previous (averagingWindow - 1) measurements.

TODOs:
1. Compute final reading with this.
2. Change type of values to float for averages.

### CODE REVIEW

**Comment by kamatajinkya** [comment]
Floating Point Unit is not available on atmega devices. Best stay from using "float" as a datatype unless absolutely needed. I have seen some nasty rounding and precision errors. Try using Fixed Point operations.

**Comment by krishnadurai** [comment]
@kamatajinkya thanks for mentioning. This implementation does not have floating points and instead does ops on int type. Please take a look.

**Comment by kamatajinkya** [comment]
Lol. Should have look at it before. Btw, not sure if the "new" keyword is supported by arduino. From what I understand, you need an OS to support that which is not a part of avr-libc C. If arduino libraries have something hidden within then not sure. Instead you can use a macro to preallocate memory at compile time.

File: TempSensor.ino [hunk 1]
```diff
@@ -142,27 +142,102 @@ uint16_t MLX90614::read16(uint8_t a) {
...
+
+class MovingAverage {
+  private:
+  	int * m_values;
+    long m_sum = 0;
+    RotatingCounter * m_rotating_counter;
+  public:
+    MovingAverage() {
+      m_values = new int[averagingWindow];
+      m_rotating_counter = new RotatingCounter(10);
```
Comment (yashtambi) [inline TempSensor.ino]: I'd suggest not using the 'new' keyword, and using a static memory allocation in the form of arrays instead.

File: TempSensor.ino [hunk 2]
```diff
@@ -142,27 +142,102 @@ uint16_t MLX90614::read16(uint8_t a) {
...
   // set up the LCD's number of columns and rows:
-  lcd.begin(16, 2);
+  lcd.begin(16, 2); 
   lcd.home ();
 }
 
+
+// TODO(krishnadurai): Temperature readings and MovingAverages
+// are typed in `int`. We may need to move to a `float` based type.
 void loop() {
```
Comment (yashtambi) [inline TempSensor.ino:226]: Need to incorporate the trigger method, but I guess you've left it for next code version?

Comment (krishnadurai) [reply TempSensor.ino:226]: Yes I'm going to raise a separate PR for it tonight.

**Comment by krishnadurai** [comment]
@yashtambi as discussed, I've added a feature flag to disable the moving average behaviour.

Could you PTAL and merge this one?

File: TempSensor.ino (hunk 1 above)
Comment (krishnadurai) [reply TempSensor.ino]: Sure I'll address this one.

File: TempSensor.ino [hunk 3]
```diff
@@ -142,27 +144,117 @@ uint16_t MLX90614::read16(uint8_t a) {
...
-  delay(500);
+  delay(refreshTempReading);
+
+  // Get temperature readings
+  int temp_celcius = mlx.readObjectTempC();
+  int temp_fahrenheit = mlx.readObjectTempF();
+
+  // Populate temperature values for moving average
+  if (flagMovingAverageEnable == true) {
+    avg_celsius.pushValue(temp_celcius);
```
Comment (kamatajinkya) [inline TempSensor.ino:237]: Clear moving average filter. You dont want the reading from the previous person to affect the reading of the current person

File: TempSensor.ino [hunk 4]
```diff
@@ -142,27 +144,117 @@ uint16_t MLX90614::read16(uint8_t a) {
...
+MovingAverage avg_fahrenheit;
+
 void loop() {
   // Turn off the display:
   lcd.noDisplay();
-  delay(500);
+  delay(refreshTempReading);
+
+  // Get temperature readings
+  int temp_celcius = mlx.readObjectTempC();
```
Comment (kamatajinkya) [inline TempSensor.ino:232]: Run this in a loop collect some fixed number of temperature readings then proceed to filter and then display

**Review by kamatajinkya — DISMISSED** [review]
Looks like you are reading just a single value and then displaying it. Instead gather a bunch of values filter them to obtain a single result and then display it and hold for some time (we can later tie it to the trigger button if needed). Else the displayed value will constantly change, and due to persistence of vision will be unreadable

TLDR: Decouple ur display and sampling rates

File: TempSensor.ino (hunk 3 above)
Comment (krishnadurai) [reply TempSensor.ino:237]: @kamatajinkya I've filed an [issue](https://github.com/CaringIndians/thermalsensor/issues/8) to resolve this comment and this is a bug. I'll address this in the [subsequent PR for trigger implementation](https://github.com/CaringIndians/thermalsensor/pull/6).

**Comment by krishnadurai** [comment]
@kamatajinkya thanks for your review and I've filed #7 to implement an aggregated filter for finalizing the temperature to show. Let us not make it a part of this PR since this is blocking the [trigger implementation PR](https://github.com/CaringIndians/thermalsensor/pull/6).

I've filed #8 and fix it in the trigger PR.

@kamatajinkya it will be safe for us to move ahead, check the [comment](https://github.com/CaringIndians/thermalsensor/issues/7#issuecomment-605444306) for the flag which disables MovingAverages.

@yashtambi it will be great if this PR can be released for #6 to be integrated with suggested fixes.

**Review by yashtambi — APPROVED** [review]

**Review by yashtambi — APPROVED** [review]
